posix: device_io: implement fdopen(), fileno(), pread(), pselect(), pwrite()

cfriedt opened the pull request:
Implement the remaining functions in the `POSIX_DEVICE_IO` Option Group.

This Option Group is requried for the PSE51, PSE52, PSE53, and PSE54 subprofiles and likely others.

[Doc Preview](https://builds.zephyrproject.io/zephyr/pr/73978/docs/services/portability/posix/index.html)

Fixes #24966
Fixes #66932 
Fixes #66938 
Fixes #66946 
Fixes #66947 
Fixes #66948

Commit:
doc: posix: mark posix device io as complete

Mark the POSIX_DEVICE_IO Option Group as commplete. This Option
Group is required for PSE51, PSE52, PSE53, PSE54, and most of
the interesting applications involving file descriptors and
sockets.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
This one has a false positive regarding the reserved word "remove" due to the `remove()` function existing in a file that was touched.

No changes made there, although it is an ISO C function, and not a POSIX one. That would be a good bugfix at some point.

Commit:
posix: device_io: require a full libc for c89 functions

The POSIX_DEVICE_IO Option Group requires a number of c89
functions mainly from stdio.h .

Namely,

clearerr(), fclose(), feof(), ferror(), fflush(), fetc(),
fgets(), fprintf(), fputc(), fputs(), fread(), freopen(),
fscanf(), fwrite(), getc(), getchar(), gets(), perror(),
printf(), putc(), putchar(), puts(), scanf(), setbuf(),
setvbuf(), ungetc(), vfprintf(), vfscanf(), vfprintf(), and
vscanf().

These should be provided by any conformant C library
(not by the POSIX API).

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: device_io: provide stdin, stdout, stderr variables

Create a compatibility Kconfig option for libc's that provide
their own stdin, stdout, stderr variables.

These are POSIX variables, so eventually we may want to
consider defaulting this option to 'y' and maybe providing
some zvfs abstraction for them.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
fdtable: read, write, close: only execute methods if non-NULL

Only invoke vtable methods read, write, and close if they are
non-NULL.

The close() vtable method is optional, so that should not return
an error if zvfs_close() is called and that method is
unimplemented.

Otherwise, if zvfs_read() or zvfs_write() are called and the
corresponding vtable method is unimplemented, fail setting
errno to EIO.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: device_io: implement pread() and pwrite()

Add pread() and pwrite() implementations, which are nearly
identical to read() and write() but differ in that they do not
update the file-descriptor offset and instead read from a
specific file offset.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
net: sockets: move select() implementation to zvfs

Move the implementation of zsock_select() to zvfs_select(). This
allows other types of file descriptors to also make use of
select() functionality even when the network subsystem is not
enabled.

Additionally, it partially removes a dependency cycle between
posix and networking by moving functionality into a mutual
dependency.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
net: sockets: move poll implementation to zvfs

Move the implementation of zsock_poll to zvfs_poll. This allows
other types of file descriptors to also make use of poll()
functionality even when the network subsystem is not enabled.

Additionally, it partially removes a dependency cycle between
posix and networking by moving functionality into a mutual
dependency.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: device_io: implement pselect()

Implement pselect() as it's required by POSIX_DEVICE_IO

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: device_io: implement fdopen()

Implement fdopen(), as required by the POSIX_DEVICE_IO Option
Group.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: device_io: implement fileno()

Implement fileno() as required by the POSIX_DEVICE_IO Option
Group.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: device_io: use mode argument correctly in open()

Previously, we had only used the flags field and ignored mode
with the open() function.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
Fixes for a few bugs prepended.

aescolar commented:
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

jukkar commented on lib/posix/options/Kconfig.device_io:
```diff
@@ -42,6 +42,14 @@ config POSIX_DEVICE_IO_ALIAS_WRITE
+config POSIX_DEVICE_IO_STDIN_STDOUT_STDERR
+	bool
+	help
+	  Select 'y' here and Zephyr will provide the stdin, stdout, and stderr variables.
```
The help text is slightly misleading as user is not able to y this option because it is hidden.

cfriedt replied:
Will fix after the TSC meeting tomorrow.

cfriedt replied:
Should be fixed now. I applied the same pattern to other non-user-selectable Kconfig as well.

ycsin commented on lib/os/fdtable.c:
```diff
@@ -72,6 +73,12 @@ static struct fd_entry fdtable[CONFIG_ZVFS_OPEN_MAX] = {
 };
 
+#ifdef CONFIG_POSIX_DEVICE_IO_STDIN_STDOUT_STDERR
+FILE *stdin = &fdtable[0];
```
I guess we might need to add `FILE` to https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/checkpatch/typedefsfile

cfriedt replied:
Maybe - I thought it might have been because the minimal libc defines `stdin`, `stdout`, `stderr` as macros. Hard to say.

ycsin replied:
It seems to fix the issue in vscode locally, created #74507

cfriedt commented on subsys/net/lib/ptp/clock.c:
```diff
@@ -55,7 +55,7 @@ struct ptp_clock {
 };
 
-static struct ptp_clock clock = { 0 };
+__maybe_unused static struct ptp_clock ptp_clk = { 0 };
```
Yes, there is a separate PR for it, but I needed to have it here too because I found that otherwise there would be a compile error.

cfriedt commented:
temporary DNM until merge conflicts and all comments have been addressed.

Commit:
posix: kconfig: remove select y from non-user-selectable help

Several help prompts for non-user-selectable Kconfig options
included the phrase "select 'y' here", which does not make
any sense in this situation.

Adjust the help sections to use more appropriate language.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
* rebased
* fixed merge conflicts
* updated language for non-user-selectable Kconfig options (i.e. do not use "Select 'y'" if the user cannot select y)

cfriedt commented:
Other than the false positive in coding guidelines, this should probably be all green. Please revisit when you have a moment @ycsin, @jukkar, @SeppoTakalo
